File: controller/tasks/task-controller.go
package tasks

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/manabie-com/backend/entity"
	"github.com/manabie-com/backend/repository"
	taskservice "github.com/manabie-com/backend/services/task"
	"github.com/manabie-com/backend/utils"

	userServiceValidate "github.com/manabie-com/backend/services/user"
)

type Controller struct {
	Service taskservice.I_TaskService
}

type I_TaskController interface {
	GetTaskAll(c *gin.Context)
	CreateTask(c *gin.Context)
	UpdateTask(c *gin.Context)
	DeleteTask(c *gin.Context)
}

func NewTaskController(serv taskservice.I_TaskService) I_TaskController {

	return &Controller{
		Service: serv,
	}
}

func (ct *Controller) GetTaskAll(c *gin.Context) {
	tasks, err := ct.Service.GetTaskAll()
	if err != nil {
		c.JSON(http.StatusInternalServerError, err)
		return
	}

	c.JSON(http.StatusOK, gin.H{"data": tasks})
}

func (ct *Controller) UpdateTask(c *gin.Context) {
	repo := repository.NewRepository()
	taskservice := taskservice.NewTaskService(repo)

	id := c.Param("id")
	var task entity.Task
	if err := c.ShouldBindJSON(&task); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"data": utils.ErrBadRequest("Invalid body json")})
		return
	}

	task.ID = id

	result, err := taskservice.UpdateTask(&task)
	if err != nil {
		c.JSON(err.Code, gin.H{"data": err})
		return
	}
	c.JSON(http.StatusOK, gin.H{"data": result})

}

func (ct *Controller) DeleteTask(c *gin.Context) {
	repo := repository.NewRepository()
	taskservice := taskservice.NewTaskService(repo)

	id := c.Param("id")

	err := taskservice.DeleteTask(id)
	if err != nil {
		c.JSON(err.Code, gin.H{"data": err})
		return
	}
	c.JSON(http.StatusOK, gin.H{"data": "ok"})

}

func (ct *Controller) CreateTask(c *gin.Context) {
	var task entity.Task

	if err := c.ShouldBindJSON(&task); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"data": utils.ErrBadRequest("Invalid body json")})

		return
	}

	repo := repository.NewRepository()
	taskServiceValidate := taskservice.NewTaskServiceValidate(repo)

	if err := taskServiceValidate.Validate(&task); err != nil {

		c.JSON(err.Code, gin.H{"data": err})

		return
	}

	userValidate, err := userServiceValidate.NewUserServiceValidate(repo, task.UserID)
	if err != nil {

		c.JSON(err.Code, gin.H{"data": err})

		return
	}

	if err := userValidate.IsAllowedAddTask(); err != nil {

		c.JSON(err.Code, gin.H{"data": err})

		return
	}

	savedTask, savedErr := ct.Service.CreateTask(&task)

	if savedErr != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"data": err})
		return
	}

	c.JSON(http.StatusCreated, gin.H{"data": savedTask})
}

File: src/apis/controllers/task.controller.js
const httpStatus = require("http-status");
const { get } = require("lodash");

const catchAsync = require("../../utils/catch-async");
const { taskService } = require("../services");

const createTask = catchAsync(async (req, res) => {
  const task = await taskService.createTask(req.body, req.user);
  res.status(httpStatus.CREATED).send(task);
});

const getTasks = catchAsync(async (req, res) => {
  const tasks = await taskService.getTasks(req.user._id);
  res.status(httpStatus.OK).send(tasks);
});

const getTaskById = catchAsync(async (req, res) => {
  const createdBy = get(req.user, "_id", "");
  const task = await taskService.getTask(req.params.id, createdBy);
  res.status(httpStatus.OK).send(task);
});

const updateTask = catchAsync(async (req, res) => {
  const createdBy = get(req.user, "_id", "");

  const task = await taskService.updateTask(req.params.id, req.body, createdBy);
  res.status(httpStatus.OK).send(task);
});

const deleteTask = catchAsync(async (req, res) => {
  const createdBy = get(req.user, "_id", "");

  await taskService.deleteTask(req.params.id, createdBy);
  res.status(httpStatus.NO_CONTENT).send();
});

module.exports = {
  createTask,
  getTasks,
  getTaskById,
  updateTask,
  deleteTask,
};
